src/binding-hash.cc

```diff
@@ -16,164 +16,77 @@
 
 #include "binding-hash.h"
 
-#define INITIAL_HASH_CAPACITY 8
+#include <algorithm>
+#include <vector>
 
 namespace wabt {
 
-static size_t hash_name(const StringSlice* name) {
-  // FNV-1a hash
-  const uint32_t fnv_prime = 0x01000193;
-  const uint8_t* bp = reinterpret_cast<const uint8_t*>(name->start);
-  const uint8_t* be = bp + name->length;
-  uint32_t hval = 0x811c9dc5;
-  while (bp < be) {
-    hval ^= static_cast<uint32_t>(*bp++);
-    hval *= fnv_prime;
-  }
-  return hval;
-}
-
-static BindingHashEntry* hash_main_entry(const BindingHash* hash,
-                                         const StringSlice* name) {
-  return &hash->entries.data[hash_name(name) % hash->entries.capacity];
-}
-
-bool hash_entry_is_free(const BindingHashEntry* entry) {
-  return !entry->binding.name.start;
-}
-
-static BindingHashEntry* hash_new_entry(BindingHash* hash,
-                                        const StringSlice* name) {
-  BindingHashEntry* entry = hash_main_entry(hash, name);
-  if (!hash_entry_is_free(entry)) {
-    assert(hash->free_head);
-    BindingHashEntry* free_entry = hash->free_head;
-    hash->free_head = free_entry->next;
-    if (free_entry->next)
-      free_entry->next->prev = nullptr;
-
-    /* our main position is already claimed. Check to see if the entry in that
-     * position is in its main position */
-    BindingHashEntry* other_entry = hash_main_entry(hash, &entry->binding.name);
-    if (other_entry == entry) {
-      /* yes, so add this new entry to the chain, even if it is already there */
-      /* add as the second entry in the chain */
-      free_entry->next = entry->next;
-      entry->next = free_entry;
-      entry = free_entry;
+namespace {
+
+typedef std::vector<const BindingHash::value_type*> BindingHashValueTypeVector;
+
+BindingHashValueTypeVector create_binding_duplicates_vector(
```

Comment:
Why not have this get a (reference or pointer) argument that it initializes, so that it doesn't have to copy the local variable "duplicates" back to the caller.